### ISSUE TITLE
Test logs get mixed across test cases

### ISSUE THREAD

**Issue by AArnott** [xunit/xunit#2377]
## Description

`dotnet test` (and Test Explorer in VS) both collect messages logged by tests to the wrong tests.
I am using xunit's `ITestOutputHelper` to log messages, and yet these messages which are meant to be collected under "Standard Output" on a per-test basis are getting mixed up, which makes analyzing test failures very difficult or impossible.

It seems to depend on logging taking place *after* the test class's DisposeAsync method completes, such that test 1 is actually logging after it allegedly finished, thus overlapping into test 2's time slice.
That said, since concurrent test execution is supported, it seems that logging is already distinguishable between concurrent tests. So why do the logs get mixed up when one test logs after its disposal?

The tested product may have async shutdown behavior that the test's DisposeAsync method cannot wait on, which is why logging happens beyond the DisposeAsync method. But I would like to capture such logs (if the test runner is still running) and associate them with the right test.

We decided this is an xunit bug rather than in the VSTest platform because we found evidence that the corruption happens within `Xunit.Sdk.TestOutputHelper`, and by the time the xunit.runner.visualstudio library hands off to the VS Test platform with the test result, the corruption has already occurred.
![image](https://user-images.githubusercontent.com/3548/132761719-190b4882-95de-4367-9340-981d62fe53ba.png)

## Steps to reproduce

Take this minimal repro:
[TestProject1.zip](https://github.com/microsoft/vstest/files/7138852/TestProject1.zip)

```
dotnet test -d log.txt -l "trx;LogFileName=console.trx"
```

Open the `console.trx` file in VS.
Go through each of the 5 tests to look at their logged messages.
Each message is prefixed with a unique test ID (and elapsed time) to record which test invocation it came from.

## Expected behavior

Each of the tests have their own logged messages. This would manifest as each message starting with the same number as every other line.

## Actual behavior

I see some test logs include *multiple* of these unique ID prefixes:
![image](https://user-images.githubusercontent.com/3548/132740977-bceeb69f-0bbf-4067-a7b1-39c4b1e9eb51.png)

It also reproduces when tests are run in Test Explorer:
![image](https://user-images.githubusercontent.com/3548/132741020-25a90aef-2e2f-47ea-9467-887a90d2c878.png)

## Diagnostic logs
[logs.zip](https://github.com/microsoft/vstest/files/7138884/logs.zip)

## Environment

Windows
.NET SDK 5.0.400

**Comment by drvic10k** [comment]
why was this issue closed? was it fixed?

**Comment by AArnott** [comment]
@drvic10k Why do you say it was closed? It's active now, and the history on the issue doesn't suggest it was ever closed. Perhaps you were confused by the link from a closed issue that appears in this issue's history?

**Comment by drvic10k** [comment]
@AArnott you are right, that is indeed a label from the linked issue

**Comment by slawomirbrys** [comment]
Hi @AArnott, 
Do you know if there is any progress on that bug?
I've just encountered the same bug on my local tests. I tried as much as possible to have concrete logs per my test (that is quite extensive in logging) and now seems that everything is mixed up 🥴. It will be very hard to quickly diagnose issues with tests on CI/CD until someone will run single test on local machine and have proper logs.

**Comment by slawomirbrys** [comment]
Ok, thank you.

**Comment by bradwilson** [comment]
This bug comes because we resolve class constructor arguments once per class, so there is a shared instance of the test output helper being used for all tests in the same class:

https://github.com/xunit/xunit/blob/e022e5cf44943fede7c616eb482f5b9d7b98bd9c/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestClassRunner.cs#L226-L230

It gets emptied and reset between tests:

https://github.com/xunit/xunit/blob/e022e5cf44943fede7c616eb482f5b9d7b98bd9c/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestRunner.cs#L64-L73

By using it outside of its intended scope, you're writing to another test, since we're reusing the instance.

What would you prefer be done here @AArnott?

- Ignore the output that's done beyond scope
- Report the attempt to write output beyond scope as a fatal error

**Comment by bradwilson** [comment]
I should point out that in v3 this is already fixed and it takes the "ignore" stance.

**Comment by bradwilson** [comment]
No dispose messages anywhere (this is your test project, updated to v3)


https://github.com/xunit/xunit/assets/16944/2d7a60d9-0db0-44db-b277-86389894710b

**Comment by bradwilson** [comment]
The "ignore" behavior is available in `2.8.2-pre.15` https://xunit.net/docs/using-ci-builds

To be clear, your call to logger.WriteLine() will now be throwing, but since nobody is observing the task you spun off, the exception will be ignored (and the test will continue to pass):

https://github.com/xunit/xunit/assets/16944/a87e26d2-2e8b-431e-b641-e9178e4477dc

**Comment by AArnott** [comment]
Great. Matching behavior between v2 and v3 seems like goodness. 

Between silently failing and throwing, I'd prefer throwing so that the problem will usually be discoverable sooner than when we have some _other_ test failure and we have to search the logs only to find that they're incomplete.

**Comment by bradwilson** [comment]
I'll add a new issue to make sure all the runners are subscribed to `AppDomain.CurrentDomain.UnhandledException`.
